src/main/java/com/google/sps/servlets/TranslateLocationServlet.java

```diff
@@ -0,0 +1,96 @@
+import com.google.cloud.secretmanager.v1.SecretVersionName;
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.maps.GaeRequestHandler;
+import com.google.maps.GeoApiContext;
+import com.google.maps.GeocodingApi;
+import com.google.maps.model.GeocodingResult;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that translates locations written in words to latitude and longitude. */
+@WebServlet("/translateLocation")
+public class TranslateLocationServlet extends HttpServlet {
+
+  /** POST a location and get back a latitude and longitude. */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Get the input from the form.
+    String location = getParameter(request, "location", "");
+
+    // Check for validity.
+    if (location.isEmpty()) {
+      response.setStatus(400);
+      return;
+    }
+
+    String projectId = "step186-2020";
+    String secretId = "geocoding-api-key";
+    String versionId = "latest";
+    String geocodingApiKey = "";
+
+    // Retrieve the secret key for the Geocoding API.
+    try (SecretManagerServiceClient client = SecretManagerServiceClient.create()) {
+      SecretVersionName secretVersionName = SecretVersionName.of(projectId, secretId, versionId);
+      AccessSecretVersionResponse secretResponse = client.accessSecretVersion(secretVersionName);
+      geocodingApiKey = secretResponse.getPayload().getData().toStringUtf8();
+    }
+
+    // Try to get the latitude and longitude for the given address.
+    try {
+      GeoApiContext context = new GeoApiContext.Builder(new GaeRequestHandler.Builder()).apiKey(geocodingApiKey).build();
+      GeocodingResult[] results =  GeocodingApi.geocode(context, location).await();
+      Gson gson = new GsonBuilder().setPrettyPrinting().create();
+
+      response.setStatus(200);
+      response.setContentType("text/json; charset=UTF-8");
+      response.setCharacterEncoding("UTF-8");
+      if (results.length > 0) {
+        response.getWriter().println(gson.toJson(results[0]));
+      }
+      else {
+        response.getWriter().println(gson.toJson(new Object()));
+      }
+    }
+    catch (Exception e) {
+      response.setStatus(500);
+      response.setContentType("text; charset=UTF-8");
+      response.getWriter().println(e.toString());
+      return;
+    }
+  }
+
+  /**
+   * @return the request parameter, or the default value if the parameter was not specified by the
+   *     client.
+   */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null || value.isEmpty()) {
+      return defaultValue;
+    }
+    return value;
+  }
+}
```

Comment:
It's recommended to end files with newlines!

Reply:
ok! both of my files have newlines locally, I'm wondering if I messed up committing them somehow